src/us/kbase/workspace/database/mongo/ObjectLister.java

```diff
@@ -119,39 +134,52 @@ private DBObject buildProjection(final ResolvedListObjectParameters params) {
 	 */
 	private DBObject buildSortSpec(final ResolvedListObjectParameters params) {
 		final DBObject sort = new BasicDBObject();
-		if (isObjectIDFiltersOnly(params)) {
+		if (isSafeForUPASort(params)) {
+			if (params.getType().isPresent()) {
+				final TypeDefId t = params.getType().get();
+				// TODO CODE use optionals for typedefid versions
+				if (t.getMinorVersion() != null) {
+					sort.put(Fields.VER_TYPE_FULL, 1);
+				} else if (t.getMajorVersion() != null) {
+					sort.put(Fields.VER_TYPE_WITH_MAJOR_VERSION, 1);
+				} else {
+					sort.put(Fields.VER_TYPE_NAME, 1);
+				}
+			}
 			sort.put(Fields.VER_WS_ID, 1);
 			sort.put(Fields.VER_ID, 1);
 			sort.put(Fields.VER_VER, -1);
 		}
 		return sort;
 	}
 	
-	/* Check if there are no filters set other than the object ID filters. The object ID filters
-	 * may or may not be set.
-	 * The other filters are the two date filters, the metadata filter, the savers filter, and
-	 * the type filter.
+	/* Check if there are no filters set other than the object ID and type filters. The object ID
+	 * and type filters may or may not be set.
+	 * The other filters are the two date filters, the metadata filter, and the savers filter.
 	 * @return true if no filters other than the object ID filters are set.
 	 */
-	private boolean isObjectIDFiltersOnly(final ResolvedListObjectParameters params) {
+	private boolean isSafeForUPASort(final ResolvedListObjectParameters params) {
 		// be really careful about modifying this function. See notes in this file.
-		boolean oidFiltersOnly = !params.getAfter().isPresent()
+		return !params.getAfter().isPresent()
 				&& !params.getBefore().isPresent()
 				&& params.getMetadata().isEmpty()
-				&& params.getSavers().isEmpty()
-				// must have workspaces specified
-				&& !params.getType().isPresent();
-		return oidFiltersOnly;
+				&& params.getSavers().isEmpty();
 	}
 
 	private DBObject buildQuery(final ResolvedListObjectParameters params) {
-		final Set<Long> ids = params.getPermissionSet().getWorkspaces().stream()
-				.map(ws -> ws.getID()).collect(Collectors.toSet());
+		final List<Long> ids = params.getPermissionSet().getWorkspaces().stream()
+				.map(ws -> ws.getID()).distinct().sorted().collect(Collectors.toList());
 		final DBObject verq = new BasicDBObject();
 		verq.put(Fields.VER_WS_ID, new BasicDBObject("$in", ids));
 		if (params.getType().isPresent()) {
-			verq.put(Fields.VER_TYPE_FULL, new BasicDBObject(
-					"$regex", "^" + params.getType().get().getTypePrefix()));
+			final TypeDefId t = params.getType().get();
+			if (t.getMinorVersion() != null) {
+				verq.put(Fields.VER_TYPE_FULL, t.getTypeString());
+			} else if (t.getMajorVersion() != null) {
+				verq.put(Fields.VER_TYPE_WITH_MAJOR_VERSION, t.getTypeString());
+			} else {
+				verq.put(Fields.VER_TYPE_NAME, t.getTypeString());
+			}
```

Comment:
since this is very similar to the code around line 137, is it worth abstracting it out into a function?

Reply:
Done